openmicroscopy/bioformats
# maven: Package bftools, bfmatlab and bioformats-octave

# Background

This is part of the decoupling/maven migration work.  I looked at this briefly due to the recent ant build breakage in case ant was going to be hard or impossible to fix.  The maven build didn't previously build the `bftools`, `bfmatlab` or `bioformats-octave` distributions.  This PR adds these three packages to the build, and will allow releases to be made with maven instead of ant.

Note that as well as the `bftools` distribution, this change also allows running of the tools like `showinf` and `bfconvert` *inside the source tree* when building with maven.  This was a critical missing piece which required building with ant in order to use them for routine development and testing.  It's now possible to use maven for most testing, removing most of the need for ant.  All the usual scripts like `showinf` and `bfconvert` will run with the correct classpath and work properly.  Likewise `ijview`, which will report the correct place to copy `ij` jar both in-tree and in bftools when built with ant or maven.  References to `loci_tools` and other jars in the `bf.sh`, `bf.bat` and other scripts have been dropped; only `bioformats_package` is bundled, making this unused logic pointless both in-tree and inside bftools.

References to the old `omeul` utility still present in the bftools packaging logic and docs have been dropped.

# Building

Running `mvn` will create the packages in `components/bundles/bioformats_package/target` (bftools, matlab and octave).  Note that the maven-produced bftools can be created on Unix or Windows and will have the correct file permissions.

Running `ant tools dist-bftools dist-matlab dist-octave` will create the ant distributions (bftools works on Unix only).  

# Testing

See build instructions above.

- Run `mvn`
- Run `ant tools dist-bftools dist-matlab dist-octave`
- Compare the contents of the maven-produced archives with the ant-produced archives.
  - The matlab and octave contents should be identical.
  - The bftools contents replace the überjar with a `jars` directory containing the individual jars, but is otherwise functionally identical.

- Run `git clean -dxf` and `mvn`
- Run `tools/showinf`, `tools/bfconvert`, `tools/ijview` inside the source tree
  - `ijview` will fail; download [ij.jar](https://imagej.nih.gov/ij/download.html) and copy to the `components/bundles/bioformats_package/target` directory
- Run `showinf`, `bfconvert` and `ijview` with the unpacked bftools zip
  - `ijview` will fail; copy `ij.jar` to the `jar` directory

- Run `git clean -dxf` and `ant tools dist-bftools dist-matlab dist-octave`
- Run `tools/showinf`, `tools/bfconvert`, `tools/ijview` inside the source tree
  - `ijview` will fail; download [ij.jar](https://imagej.nih.gov/ij/download.html) and copy to the `artifacts` directory
- Run `showinf`, `bfconvert` and `ijview` with the unpacked bftools zip
  - `ijview` will fail; copy `ij.jar` to the root of the unpacked zip

I have tested these combinations:

Build | Test in tree | Test zip
--------|----------|---------
mvn (Unix) | showinf/ijview/ijview+ij.jar (Unix, Win) | showinf/ijview/ijview+ij.jar (Unix, Win)
ant tools dist-bftools (Unix) | showinf/ijview/ijview+ij.jar (Unix, Win) | showinf/ijview/ijview+ij.jar (Unix, Win)
mvn (Win) | showinf/ijview/ijview+ij.jar (Unix, Win) | showinf/ijview/ijview+ij.jar (Unix, Win)
ant tools (Win) | showinf/ijview/ijview+ij.jar (Unix, Win) | N/A (bftools.jar only created on Unix)

Note ijview was tested with and without the presence of the `ij` jar to ensure it failed correctly.

The size of this test matrix is a large reason why we should be looking at dropping the `ant` build entirely!
